### PR TITLE
feat(amf): Serialization and Deserialization of Smf Context

### LINKED ISSUE #10924: smf context Serialize and Deserialize

### PR DESCRIPTION
**Pull request by shashidhar-patil**
## Summary
**Serialization and Deserialization of Smf Context to support stateless feature.**

## Test Plan
- Verified basic sanity with ueransim
- Verified using Functional UT
- Please find attached logs and pcap for basic registration

MME logs:
[mme.log](https://github.com/magma/magma/files/7760565/mme.log)

pcap snapshot:
smf_context_pcap

Unit test Snapshot:
smf_context_UT

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp [hunk 1]
```diff
@@ -208,4 +208,296 @@ void AmfNasStateConverter::proto_to_ue_m5gmm_context(
...
+    const magma::lte::oai::Ambr& ambr_proto,
+    session_ambr_t* state_session_ambr) {
+  state_session_ambr->dl_ambr_unit    = ambr_proto.br_unit();
+  state_session_ambr->dl_session_ambr = ambr_proto.br_dl();
+  state_session_ambr->ul_ambr_unit    = ambr_proto.br_unit();
+  state_session_ambr->ul_session_ambr = ambr_proto.br_ul();
+}
+
+void AmfNasStateConverter::qos_flow_level_parameters_to_proto(
+    const qos_flow_level_qos_parameters& state_qos_flow_paramters,
```
[the remark this answers is not in the document]
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp [hunk 2]
```diff
@@ -208,4 +208,296 @@ void AmfNasStateConverter::proto_to_ue_m5gmm_context(
...
+    magma::lte::oai::QosFlowParameters* qos_flow_paramters_proto) {
+  qos_flow_paramters_proto->set_fiveqi(
```
[the remark this answers is not in the document]
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp [hunk 3]
```diff
@@ -208,4 +208,296 @@ void AmfNasStateConverter::proto_to_ue_m5gmm_context(
...
+    session_ambr_t* state_session_ambr) {
+  state_session_ambr->dl_ambr_unit    = ambr_proto.br_unit();
+  state_session_ambr->dl_session_ambr = ambr_proto.br_dl();
+  state_session_ambr->ul_ambr_unit    = ambr_proto.br_unit();
+  state_session_ambr->ul_session_ambr = ambr_proto.br_ul();
+}
+
+void AmfNasStateConverter::qos_flow_level_parameters_to_proto(
+    const qos_flow_level_qos_parameters& state_qos_flow_paramters,
+    magma::lte::oai::QosFlowParameters* qos_flow_paramters_proto) {
```
[the remark this answers is not in the document]
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp [hunk 4]
```diff
@@ -208,4 +208,296 @@ void AmfNasStateConverter::proto_to_ue_m5gmm_context(
...
+  qos_flow_paramters_proto->set_fiveqi(
+      state_qos_flow_paramters.qos_characteristic.non_dynamic_5QI_desc.fiveQI);
+  qos_flow_paramters_proto->set_priority_level(
```
[the remark this answers is not in the document]
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp [hunk 5]
```diff
@@ -208,4 +208,296 @@ void AmfNasStateConverter::proto_to_ue_m5gmm_context(
...
+      state_qos_flow_paramters.alloc_reten_priority.priority_level);
+  qos_flow_paramters_proto->set_preemption_vulnerability(
```
[the remark this answers is not in the document]
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp [hunk 6]
```diff
@@ -208,4 +208,296 @@ void AmfNasStateConverter::proto_to_ue_m5gmm_context(
...
+      state_qos_flow_paramters.alloc_reten_priority.pre_emption_vul);
+  qos_flow_paramters_proto->set_preemption_capability(
```
[the remark this answers is not in the document]
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp [hunk 7]
```diff
@@ -208,4 +208,296 @@ void AmfNasStateConverter::proto_to_ue_m5gmm_context(
...
+      state_qos_flow_paramters.alloc_reten_priority.pre_emption_cap);
+}
+
+void AmfNasStateConverter::proto_to_qos_flow_level_paramerters(
+    const magma::lte::oai::QosFlowParameters& qos_flow_paramters_proto,
+    qos_flow_level_qos_parameters* state_qos_flow_paramters) {
```
[the remark this answers is not in the document]
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp [hunk 8]
```diff
@@ -208,4 +208,296 @@ void AmfNasStateConverter::proto_to_ue_m5gmm_context(
...
+  qos_flow_paramters_proto->set_priority_level(
+      state_qos_flow_paramters.alloc_reten_priority.priority_level);
+  qos_flow_paramters_proto->set_preemption_vulnerability(
+      state_qos_flow_paramters.alloc_reten_priority.pre_emption_vul);
+  qos_flow_paramters_proto->set_preemption_capability(
+      state_qos_flow_paramters.alloc_reten_priority.pre_emption_cap);
+}
+
+void AmfNasStateConverter::proto_to_qos_flow_level_paramerters(
+    const magma::lte::oai::QosFlowParameters& qos_flow_paramters_proto,
```
[the remark this answers is not in the document]
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp [hunk 9]
```diff
@@ -208,4 +208,296 @@ void AmfNasStateConverter::proto_to_ue_m5gmm_context(
...
+    qos_flow_level_qos_parameters* state_qos_flow_paramters) {
+  state_qos_flow_paramters->qos_characteristic.non_dynamic_5QI_desc.fiveQI =
```
[the remark this answers is not in the document]
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp [hunk 10]
```diff
@@ -208,4 +208,296 @@ void AmfNasStateConverter::proto_to_ue_m5gmm_context(
...
+      qos_flow_paramters_proto.fiveqi();
```
[the remark this answers is not in the document]
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp [hunk 11]
```diff
@@ -208,4 +208,296 @@ void AmfNasStateConverter::proto_to_ue_m5gmm_context(
...
+  state_qos_flow_paramters->alloc_reten_priority.priority_level =
```
[the remark this answers is not in the document]
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp [hunk 12]
```diff
@@ -208,4 +208,296 @@ void AmfNasStateConverter::proto_to_ue_m5gmm_context(
...
+      qos_flow_paramters_proto.priority_level();
```
[the remark this answers is not in the document]
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp [hunk 13]
```diff
@@ -208,4 +208,296 @@ void AmfNasStateConverter::proto_to_ue_m5gmm_context(
...
+  state_qos_flow_paramters->alloc_reten_priority.pre_emption_vul =
```
[the remark this answers is not in the document]
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp [hunk 14]
```diff
@@ -208,4 +208,296 @@ void AmfNasStateConverter::proto_to_ue_m5gmm_context(
...
+      static_cast<pre_emption_vulnerability>(
+          qos_flow_paramters_proto.preemption_vulnerability());
```
[the remark this answers is not in the document]
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp [hunk 15]
```diff
@@ -208,4 +208,296 @@ void AmfNasStateConverter::proto_to_ue_m5gmm_context(
...
+  state_qos_flow_paramters->alloc_reten_priority.pre_emption_cap =
```
[the remark this answers is not in the document]
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp [hunk 16]
```diff
@@ -208,4 +208,296 @@ void AmfNasStateConverter::proto_to_ue_m5gmm_context(
...
+      static_cast<pre_emption_capability>(
+          qos_flow_paramters_proto.preemption_capability());
```
[the remark this answers is not in the document]
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.h [hunk 17]
```diff
@@ -60,6 +60,73 @@ class AmfNasStateConverter : public magma::lte::StateConverter {
...
+  static void qos_flow_setup_request_item_to_proto(
+      const qos_flow_setup_request_item& state_qos_flow_request_item,
+      magma::lte::oai::M5GQosFlowItem* qos_flow_item_proto);
+
+  static void proto_to_qos_flow_setup_request_item(
+      const magma::lte::oai::M5GQosFlowItem& qos_flow_item_proto,
+      qos_flow_setup_request_item* state_qos_flow_request_item);
+
+  static void qos_flow_level_parameters_to_proto(
+      const qos_flow_level_qos_parameters& state_qos_flow_paramters,
```
[the remark this answers is not in the document]
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.h]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.h [hunk 18]
```diff
@@ -60,6 +60,73 @@ class AmfNasStateConverter : public magma::lte::StateConverter {
...
+      magma::lte::oai::QosFlowParameters* qos_flow_paramters_proto);
```
[the remark this answers is not in the document]
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.h]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.h [hunk 19]
```diff
@@ -60,6 +60,73 @@ class AmfNasStateConverter : public magma::lte::StateConverter {
...
+
+  static void proto_to_qos_flow_level_paramerters(
+      const magma::lte::oai::QosFlowParameters& qos_flow_paramters_proto,
```
[the remark this answers is not in the document]
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.h]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.h [hunk 20]
```diff
@@ -60,6 +60,73 @@ class AmfNasStateConverter : public magma::lte::StateConverter {
...
+      qos_flow_level_qos_parameters* state_qos_flow_paramters);
```
[the remark this answers is not in the document]
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.h]: Resolved.

**Comment by shashidhar-patil** [comment]
**PR Rebased.**

mme log: [05/01/2022]
[mme.log](https://github.com/magma/magma/files/7812878/mme.log)

pcap snap: [05/01/2022]
smfcontext_pcap

Unit test: [05/01/2022]
smf_context_2

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp [hunk 21]
```diff
@@ -208,4 +208,301 @@ void AmfNasStateConverter::proto_to_ue_m5gmm_context(
...
+  state_smf_proc_data->max_downlink = smf_proc_data_proto.max_downlink();
+  state_smf_proc_data->pdu_session_type =
+      static_cast<M5GPduSessionType>(smf_proc_data_proto.pdu_session_type());
+  state_smf_proc_data->ssc_mode = smf_proc_data_proto.ssc_mode();
+}
+
+void AmfNasStateConverter::s_nssai_to_proto(
+    const s_nssai_t* state_s_nssai, magma::lte::oai::SNssai* snassi_proto) {
+  snassi_proto->set_sst(state_s_nssai->sst);
+  snassi_proto->set_sd(*(uint32_t*) &state_s_nssai->sd);
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: No need to convert to pointer and then dereference.

```suggestion
  snassi_proto->set_sd(state_s_nssai->sd);
```

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp [hunk 22]
```diff
@@ -208,4 +208,301 @@ void AmfNasStateConverter::proto_to_ue_m5gmm_context(
...
+}
+void AmfNasStateConverter::proto_to_s_nssai(
+    const magma::lte::oai::SNssai& snassi_proto, s_nssai_t* state_s_nssai) {
+  state_s_nssai->sst              = snassi_proto.sst();
+  *(uint32_t*) &state_s_nssai->sd = snassi_proto.sd();
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]:
```suggestion
state_s_nssai->sd = snassi_proto.sd();
```

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp [hunk 23]
```diff
@@ -208,4 +208,301 @@ void AmfNasStateConverter::proto_to_ue_m5gmm_context(
...
+      state_smf_context->gtp_tunnel_id.gnb_gtp_teid);
+
+  char gnb_gtp_teid_ip_addr_str[16] = {0};
+  inet_ntop(
+      AF_INET, state_smf_context->gtp_tunnel_id.gnb_gtp_teid_ip_addr,
+      gnb_gtp_teid_ip_addr_str, INET_ADDRSTRLEN);
+  smf_context_proto->set_gnb_gtp_teid_ip_addr(gnb_gtp_teid_ip_addr_str);
+
+  smf_context_proto->set_upf_gtp_teid(
+      *(uint32_t*) &state_smf_context->gtp_tunnel_id.upf_gtp_teid);
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Why do we need the pointer creation and dereferencing?
```suggestion
      state_smf_context->gtp_tunnel_id.upf_gtp_teid);
```

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp [hunk 24]
```diff
@@ -208,4 +208,301 @@ void AmfNasStateConverter::proto_to_ue_m5gmm_context(
...
+      smf_context_proto.gnb_gtp_teid();
+
+  memset(
+      &state_smf_context->gtp_tunnel_id.gnb_gtp_teid_ip_addr, '\0',
+      sizeof(state_smf_context->gtp_tunnel_id.gnb_gtp_teid_ip_addr));
+  inet_pton(
+      AF_INET, smf_context_proto.gnb_gtp_teid_ip_addr().c_str(),
+      &(state_smf_context->gtp_tunnel_id.gnb_gtp_teid_ip_addr));
+
+  *(uint32_t*) &state_smf_context->gtp_tunnel_id.upf_gtp_teid =
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]:
```suggestion
  state_smf_context->gtp_tunnel_id.upf_gtp_teid =
```

**Review by ssanadhya — CHANGES_REQUESTED** [review]

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp (hunk 21 above)
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Here "sd" is a 3-byte raw buffer (uint8_t) but we are storing it as uint32 in protobuff. So this conversion is required. We thought of a couple of approaches and implemented this.
The other approach would be to store sd as "bytes" type in protobuff. But this approach will need conversion to char and then storing it and converting it back while retrieving.
@ssanadhya Please suggest what would be the best approach to this.

Other approach pseudo code(sd is of type "bytes"):
void AmfNasStateConverter::sd_to_chars(const uint8_t* sd, char* sd_array) {
  sd_array[0] = static_cast<char>(sd[0] + ASCII_ZERO);
  sd_array[1] = static_cast<char>(sd[1] + ASCII_ZERO);
  sd_array[2] = static_cast<char>(sd[2] + ASCII_ZERO);
}

void AmfNasStateConverter::chars_to_sd(const char* sd_array, uint8_t* sd) {
  sd[0] = static_cast<int>(sd_array[0]) - ASCII_ZERO;
  sd[1] = static_cast<int>(sd_array[1]) - ASCII_ZERO;
  sd[2] = static_cast<int>(sd_array[2]) - ASCII_ZERO;
}

Comment (ssanadhya) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: @shashidhar-patil , you can use the "bytes" type in protobuf and pass the uint8_t array as char array:

State to proto
```
snassi_proto->set_sd((char*) state_s_nssai->sd, SD_LENGTH);
```

Proto to state
```
memcpy((void*)state_s_nssai->sd, (void*)snassi_proto.sd().c_str(), SD_LENGTH);
```

Note that you do not need to convert it into an ASCII value unless you want to print it as text.

Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Thankyou @ssanadhya. Addressed this review comment.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp (hunk 22 above)
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Addressed this review comment.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp (hunk 23 above)
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Addressed this review comment.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp (hunk 24 above)
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Addressed this review comment.

**Review by ssanadhya — APPROVED** [review]
LGTM

**Comment by ssanadhya** [comment]
Can be merged once @ardzoht approves.

**Review by ardzoht — APPROVED** [review]
lgtm

**Pull request merged**